### PR TITLE
Implement some basic pruning strategy

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -232,7 +242,13 @@ func (mp *MessagePool) Close() error {
 	return nil
 }
 
-func (mp *MessagePool) repubLocal() {
+func (mp *MessagePool) Prune() {
+	mp.pruneTrigger <- struct{}{}
+	mp.pruneTrigger <- struct{}{}
+	mp.pruneTrigger <- struct{}{}
```
Comment (Kubuxu) [inline chain/messagepool/messagepool.go:248]: 3x?

Comment (whyrusleeping) [reply chain/messagepool/messagepool.go:248]: so, its a single slot buffered channel. The first send fills the channel, the second send goes through when the pruning starts, and the third send goes through (and noops) after the pruning finishes and goes through the loop again

I should probably add a comment

File: chain/messagepool/pruning.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+	}
+
+	pruneCount -= npruned
+	if pruneCount <= 0 {
+		return nil
+	}
+
+	// Step 2. prune messages with the lowest gas prices
+
+	log.Warnf("still need to prune %d messages", pruneCount)
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:49]: you don't trigger `pruneMessages`.

File: chain/messagepool/pruning.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+	return nil
+}
+
+// just copied from miner/ SelectMessages
+func (mp *MessagePool) pruneMessages(ctx context.Context) error {
+	al := func(ctx context.Context, addr address.Address, tsk types.TipSetKey) (*types.Actor, error) {
+		return mp.api.StateGetActor(addr, mp.curTs)
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:56]: You need to take `curTsLk`

File: chain/messagepool/pruning.go (hunk 2 above)
Comment (arajasek) [inline chain/messagepool/pruning.go:49]: wanna spell this out with "after pruning future messages, still need to prune %d..."

Comment (arajasek) [reply chain/messagepool/pruning.go:49]: um, also wanna actually call pruneMessages

File: chain/messagepool/pruning.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+func (mp *MessagePool) pruneFutureMessages() (int, error) {
+	var pruned int
+	for addr, ms := range mp.pending {
+		if _, ok := mp.localAddrs[addr]; ok {
+			continue
+		}
+
+		act, err := mp.api.StateGetActor(addr, nil)
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:281]: this should use mpool's tipset key

File: chain/messagepool/pruning.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+				outBySender[bestSender].msgs = outBySender[bestSender].msgs[nBest:]
+				outBySender[bestSender].gasLimit = outBySender[bestSender].gasLimit[nBest:]
+				outBySender[bestSender].gasReward = outBySender[bestSender].gasReward[nBest:]
+
+				if len(outBySender[bestSender].msgs) == 0 {
+					delete(outBySender, bestSender)
+				}
+			}
+
+			if len(out) >= build.BlockMessageLimit {
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:232]: here and bit above you limit to BlockMessageLimit

File: chain/messagepool/pruning.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+				}
+			}
+
+			if nBest == 0 {
+				break // block gas limit reached
+			}
+
+			{
+				out = append(out, outBySender[bestSender].msgs[:nBest]...)
+				gasLimitLeft -= outBySender[bestSender].gasLimit[nBest-1]
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:221]: via this you limit messages to one block gas limit
we will end up reaching `gasLimitLeft ~= 0` and iterating through all senders trying to add more messages but we will always fail in for loop above

Comment (whyrusleeping) [reply chain/messagepool/pruning.go:221]: yeap, fixing now

File: chain/messagepool/pruning.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+		sm.gasReward = append(sm.gasReward, big.Add(sm.lastReward, estimatedReward))
+		sm.lastReward = sm.gasReward[len(sm.gasReward)-1]
+
+		sm.msgs = append(sm.msgs, msg)
+
+		outBySender[from] = sm
+	}
+
+	gasLimitLeft := int64(build.BlockGasLimit)
```
Comment (magik6k) [inline chain/messagepool/pruning.go:172]: Probably don't need this

File: chain/messagepool/pruning.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+						bestSender = sender
+						nBest = n + 1
+						bestGasToReward = gasToReward
+					}
+				}
+			}
+
+			if nBest == 0 {
+				break // block gas limit reached
+			}
```
Comment (Kubuxu) [inline chain/messagepool/pruning.go:199]: this seems useless

File: chain/messagepool/messagepool_test.go [hunk 9]
```diff
@@ -233,3 +238,52 @@ func TestRevertMessages(t *testing.T) {
...
+			t.Fatal(err)
+		}
+	}
+
+	for i := 10; i < 50; i++ {
+		smsg := mock.MkMessage(sender, target, uint64(i), w)
+		if err := mp.Add(smsg); err != nil {
+			t.Fatal(err)
+		}
+	}
```
Comment (magik6k) [inline chain/messagepool/messagepool_test.go:278]: Testing with
```suggestion
	for i := 0; i < 15; i++ {
		smsg := mock.MkMessage(sender, target, uint64(i), w)
		if err := mp.Add(smsg); err != nil {
			t.Fatal(err)
		}
	}

	for i := 20; i < 50; i++ {
		smsg := mock.MkMessage(sender, target, uint64(i), w)
		if err := mp.Add(smsg); err != nil {
			t.Fatal(err)
		}
	}
```

Gets me 11 messages after prune, shouldn't it give 10?

Comment (whyrusleeping) [reply chain/messagepool/messagepool_test.go:278]: that should give you 15 messages after prune

File: chain/messagepool/pruning.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+			// TODO: This is O(n^2)-ish, could use something like container/heap to cache this math
+			for _, sender := range orderedSenders {
+				meta, ok := outBySender[sender]
+				if !ok {
+					continue
+				}
+				for n := range meta.msgs {
+
+					if n+len(out) > mp.maxTxPoolSizeLo {
```
Comment (magik6k) [inline chain/messagepool/pruning.go:182]:
```suggestion
					if n+len(out) >= mp.maxTxPoolSizeLo {
```

File: chain/messagepool/pruning.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+
+	orderedSenders := make([]address.Address, 0, len(outBySender))
+	for k := range outBySender {
+		orderedSenders = append(orderedSenders, k)
+	}
+	sort.Slice(orderedSenders, func(i, j int) bool {
+		return bytes.Compare(orderedSenders[i].Bytes(), orderedSenders[j].Bytes()) == -1
+	})
+
+	out := make([]*types.SignedMessage, 0, build.BlockMessageLimit)
```
Comment (magik6k) [inline chain/messagepool/pruning.go:167]:
```suggestion
	out := make([]*types.SignedMessage, 0, mp.maxTxPoolSizeLo)
```

File: chain/messagepool/messagepool_test.go (hunk 9 above)
Comment (magik6k) [reply chain/messagepool/messagepool_test.go:278]: I'd expect 10, because `mp.maxTxPoolSizeLo = 10`

**Pull request merged**